crates/winch/src/compiler.rs

```diff
@@ -17,33 +20,162 @@ impl Compiler {
+        // share allocations. Look at the wasmtime_cranelift::Compiler to see
+        // how we can re-use existing context objects.
+        let validator = validator.into_validator(Default::default());
+
+        let buffer = isa
+            .compile_function(&sig, &body, validator)
+            .map_err(|e| CompileError::Codegen(format!("{:?}", e)))?;
+
+        let info = WasmFunctionInfo {
+            start_srcloc: FilePos::new(body.get_binary_reader().original_position() as u32),
+            stack_maps: mach_stack_maps_to_stack_maps(buffer.stack_maps()).into(),
+        };
+
+        let traps = buffer.traps().into_iter().map(mach_trap_to_trap).collect();
+
+        Ok((
+            info,
+            Box::new(CompiledFunction {
+                traps,
+                body: buffer.data().to_vec(),
+            }),
+        ))
     }
 
     fn compile_host_to_wasm_trampoline(
         &self,
-        _ty: &wasmtime_environ::WasmFuncType,
+        ty: &wasmtime_environ::WasmFuncType,
     ) -> Result<Box<dyn Any + Send>, CompileError> {
-        todo!()
+        let buffer = self
+            .isa
+            .compile_trampoline(ty)
+            .map_err(|e| CompileError::Codegen(format!("{:?}", e)))?;
+
+        Ok(Box::new(CompiledFunction {
+            traps: Vec::new(),
+            body: buffer.data().to_vec(),
+        }))
     }
 
     fn append_code(
         &self,
-        _obj: &mut Object<'static>,
-        _funcs: &[(String, Box<dyn Any + Send>)],
+        obj: &mut Object<'static>,
+        funcs: &[(String, Box<dyn Any + Send>)],
         _tunables: &Tunables,
-        _resolve_reloc: &dyn Fn(usize, FuncIndex) -> usize,
+        resolve_reloc: &dyn Fn(usize, FuncIndex) -> usize,
     ) -> Result<Vec<(SymbolId, FunctionLoc)>> {
-        todo!()
+        let mut builder = ModuleTextBuilder::new(obj, &*self.isa, funcs.len());
+        let mut traps = TrapEncodingBuilder::default();
+
+        // High level overview:
+        // Take the object that is being created. Append all compiled functions
+        // in the .text section for executable code and do a final check to make
+        // sure all the right data is in the object file. Take the traps within
+        // a function and append it to the .wasmtime.traps section.
+
+        let mut ret = Vec::with_capacity(funcs.len());
+        for (i, (sym, func)) in funcs.iter().enumerate() {
+            let func = func.downcast_ref::<CompiledFunction>().unwrap();
+
+            let (sym, range) = builder.append_func(&sym, func, |idx| resolve_reloc(i, idx));
+            traps.push(range.clone(), &func.traps);
+            let info = FunctionLoc {
+                start: u32::try_from(range.start).unwrap(),
+                length: u32::try_from(range.end - range.start).unwrap(),
+            };
+            ret.push((sym, info));
+        }
+
+        builder.finish();
+
+        traps.append_to(obj);
+
+        Ok(ret)
```

Comment:
Outline any differences between Cranelift and Winch for this function to see if we need to add anything in the future